vwifi.c

```diff
@@ -859,6 +949,49 @@ static netdev_tx_t vwifi_ndo_start_xmit(struct sk_buff *skb,
             }
         }
     }
+    /* TX by interface of AD-HOC mode */
+    else if (vif->wdev.iftype == NL80211_IFTYPE_ADHOC) {
+        /* Check if the packet is broadcasting */
+        if (is_broadcast_ether_addr(eth_hdr->h_dest)) {
+            list_for_each_entry (dest_vif, &vwifi->ibss_list, ibss_list) {
+                /* Don't send broadcast packet back to the source interface.
+                 */
+                if (ether_addr_equal(eth_hdr->h_source,
+                                     dest_vif->ndev->dev_addr))
+                    continue;
+                /* Don't send packet from dest_vif's denylist */
+                if (denylist_check(dest_vif->ndev->name, vif->ndev->name))
+                    continue;
+                /* Don't send packet to device with different SSID. */
+                if (strcmp(vif->ssid, dest_vif->ssid))
+                    continue;
+                /* Don't send packet to device with different BSSID. */
```

Comment:
TODO: move the packet filtering part on `netif_rx()`. Receivers with promiscuous mode on should expose the packets to kernel (via BPF) even if the packets are not for itself.

Reply:
Do you mean that I should rewrite `netif_rx()` separately, or perform packet filtering in `vwifi_rx()`?

Reply:
The MAC address here refers to the MAC address of the IBSS network itself, rather than the MAC address of any individual device within the IBSS network.